src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,142 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    public MyLinkedList() {
+    }
+
     @Override
     public void add(T value) {
+        Node<T> node = new Node<>(null, value, null);
+        if (head == null) {
+            head = node;
+        } else {
+            node.prev = tail;
+            tail.next = node;
+        }
+        tail = node;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index " + index + " is invalid");
+        }
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> element = findElement(index);
+            Node<T> node = new Node<>(element.prev, value, element);
+            if (element.prev == null) {
+                head = node;
+            } else {
+                element.prev.next = node;
+            }
+            element.prev = node;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> element = findElement(index);
```

Comment:
looks like we always call `checkIndex ` before `findElement `
so let's move `checkIndex` in `findElement` method